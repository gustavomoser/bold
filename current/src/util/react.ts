import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
