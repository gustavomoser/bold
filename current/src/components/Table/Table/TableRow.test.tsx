import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { Table, TableBody, TableCell, TableRow } from './index'

describe('TableRow', () => {
  it('should have the pointer class if onClick is specified', () => {
    const clickHandler = jest.fn()
    const { container } = render(
      <Table>
        <TableBody>
          <TableRow onClick={clickHandler}>
            <TableCell>Test</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    )
    expect(container).toMatchSnapshot()

    fireEvent.click(container.querySelector('tr'))
    expect(clickHandler).toHaveBeenCalled()
  })
  it('should NOT have the pointer class if onClick is not specified', () => {
    const { container } = render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>Test</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    )
    expect(container).toMatchSnapshot()
  })
  it('should accept the style prop', () => {
    const { container } = render(
      <Table>
        <TableBody>
          <TableRow style={{ color: 'red' }}>
            <TableCell>Test</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    )
    expect(container).toMatchSnapshot()
  })
})
