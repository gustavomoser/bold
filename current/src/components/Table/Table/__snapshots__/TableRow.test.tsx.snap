// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should NOT have the pointer class if onClick is not specified 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  border: 1px solid #D3D4DD;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
}

<div>
  <table
    class="emotion-3"
  >
    <tbody
      class="emotion-1"
    >
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-0"
        >
          Test
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableRow should accept the style prop 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  border: 1px solid #D3D4DD;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
}

.emotion-1 {
  color: red;
}

<div>
  <table
    class="emotion-3"
  >
    <tbody
      class="emotion-2"
    >
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-0"
        >
          Test
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableRow should have the pointer class if onClick is specified 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  border: 1px solid #D3D4DD;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
}

.emotion-1 {
  cursor: pointer;
}

<div>
  <table
    class="emotion-3"
  >
    <tbody
      class="emotion-2"
    >
      <tr
        class="emotion-1"
      >
        <td
          class="emotion-0"
        >
          Test
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
